src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PopupWindow.cs

```diff
@@ -14,38 +14,137 @@ public enum PopupViewType
             AuthenticationView
         }
 
+        [SerializeField] private bool shouldCloseOnFinish;
         [SerializeField] private PopupViewType activeViewType;
-
         [SerializeField] private AuthenticationView authenticationView;
         [SerializeField] private PublishView publishView;
         [SerializeField] private LoadingView loadingView;
 
+        [NonSerialized] private IApiClient client;
+
         public event Action<bool> OnClose;
 
         [MenuItem("GitHub/Authenticate")]
         public static void Launch()
         {
-            Open(PopupViewType.AuthenticationView);
+            OpenWindow(PopupViewType.AuthenticationView);
         }
 
-        public static PopupWindow Open(PopupViewType popupViewType, Action<bool> onClose = null)
+        public static PopupWindow OpenWindow(PopupViewType popupViewType, Action<bool> onClose = null)
         {
             var popupWindow = GetWindow<PopupWindow>(true);
 
-            popupWindow.OnClose.SafeInvoke(false);
+            popupWindow.Open(popupViewType, onClose);
+
+            return popupWindow;
+        }
+
+        private void Open(PopupViewType popupViewType, Action<bool> onClose)
+        {
+            OnClose.SafeInvoke(false);
+            OnClose = null;
+
+            Logger.Trace("OpenView: {0}", popupViewType.ToString());
+
+            var viewNeedsAuthentication = popupViewType == PopupViewType.PublishView;
+            if (viewNeedsAuthentication)
+            {
+                Logger.Trace("Validating to open view");
+
+                Client.ValidateCurrentUser(() => {
+
+                    Logger.Trace("User validated opening view");
+
+                    OpenInternal(popupViewType, onClose);
+                    shouldCloseOnFinish = true;
+
+                }, exception => {
+                    Logger.Trace("User required validation opening AuthenticationView");
+
+                    string message = null;
+                    string username = null;
+
+                    var usernameMismatchException = exception as TokenUsernameMismatchException;
+                    if (usernameMismatchException != null)
+                    {
+                        message = "Your credentials need to be refreshed";
+                        username = usernameMismatchException.CachedUsername;
+                    }
+
+                    var keychainEmptyException = exception as KeychainEmptyException;
+                    if (keychainEmptyException != null)
+                    {
+                        message = "We need you to authenticate first";
+                    }
+
+                    if (usernameMismatchException == null && keychainEmptyException == null)
+                    {
+                        message = "There was an error validating your account";
```

Comment:
All of these messages should be in const fields somewhere so we can localize them later